### DISCUSSION TITLE
@0xMantle Online hackathon

### DISCUSSION THREAD

**Discussion by Demian101** [zkp-co-learning/zkp-co-learn#72] [I've got an idea!]
## Codes Ref: 

 - 我昨天在看Tornado Cash https://github.com/tornadocash/tornado-core，这个项目很简单

 - 我最近在看的一个 zk-vote 的极简实现：
https://github.com/TheBojda/zktree-vote/tree/main/src

## Ideas：
- 我在hackerhouse线下，看到挺多idea，分享给各位：
1. zkERC20，2. zkNFTAuction，3. zkKYC, 4. zkVM, 5’ zkDSL

Zk email on chain verification 
> zkemail.xyz 有在做类似的尝试

**Comment by Demian101** [comment]
## 04/12 Meeting

#### 会议目标：

1. 现有 idea 讨论
2. zkERC20（shielded pool）  2. zkNFTAuction   3. zkKYC   4. zkVM, 5 zkDSL (类似 Circom 领域特定语言（英语：domain-specific language））
3. TornadoCash: Prove your monry is clean. - White list ...
4. 组队

#### Zk email on chain verification

类似不透露隐私的邮箱验证（@**Alan_W**）
@Abby、冯桃桃
 - 正在研究 `zkemail.xyz` 
 - 聊一些八卦（不利于大公司的一些秘密㊙️），验证邮箱证明你是企业的人，但是这一步把真正的邮箱地址隐藏掉
 - zk 版本的 “脉脉” —— teamblind 声称自己能保护用户的真正邮箱，但是不确定是不是真的。
 - Alan_W :  @ali.com  @jd.com @bytedance.com 写在 Merkle 树里面；Solidity 合约 ( 重复的邮箱验证：看一下公司在不在树里面。）
 - 邮箱前缀的字符串较少，容易暴力破解：比如 `zhangsan@bilibili.com` ，可以存储一个 cheat map ，所以必须加一些东西一起 Hash
 -  Vote ：@demian
 - 邮件去重，Group Signature，证明在 Group 里面，但是不能反推是谁
     - 没法在电路里隐藏私钥，用一个环 （门限签名）来做：
     -  我产生一个 公钥、私钥、 发给 n 个人， t 个人签名后， 把 t 个签名放在一起，组成一个完整的签名，这个签名是可以被公钥认证的。
     -  让 t / n 给我用户名的一部分（based 我是 Email 的主人）
     - 发邮件的过程需要 Server 验证（签名） 拿着这个签名去请求上面 👆🏻 的碎片。
 - @Alan Draw a picture.....
 
#### TornadoCash 优化 :
@Smart J
 - 让转账变得更加合法
 - 限制一些数额、次数？
 - 要求你的叶子节点转账固定的数量 —— 因为转账的数字不同的，可以关联出来谁是谁，比如你存入 3.4492，又存了 3.4492，可以反推。
 - @地主： 每个地址有一个转账的上限。类似国内的每人每年只能转 5W 外汇。
 - require(Mapping[address => ...] ...,  '"..."); 创新不足?

#### zk - markerdao
@H+H
我想做 zk 的 markerdao ，想通过 zk 隐藏发行稳定币后面的所有超额资产和国库资产，只给出有足够的可赎回性，防止在暴跌过程中，产生坏账，导致二级市场信心不足，稳定币被挤兑，现在难点有好多，第一，我只看过curve和uniswap v2这些dex 源码，对超额抵押稳定币代码层的逻辑不懂，第二就是想法很初级，目前个人只是会写solidity和circom，没有具体工程经验，发出来看看大家有没有觉得有意思，拉我一起讨论下，还有我觉得这个想法也太难，感觉两周做不出来

#### ZKML

 - 外包大规模计算，不揭露计算结果，证明我经过了计算。
 - 搞一个巨复杂的大电路，才能搞一下小模型
 - HunterZHunter （上一期 Hackhouse）
     - 系统提示了一个词西瓜，我要去在环境里照一个西瓜照片，然后线上找一个 ImageNet 去验证，inference (Pic into DN)，主网确定之后就可以返回一个奖励值。
     - 费了九牛二虎之力放到 智能合约里面。 
     - 如果换了计算图，就需要重新写电路。
     - inference 的计算比较贵，我证明的是：我真的经过了这个计算。 类似计算迁移。

#### ZK NFT

 - 发个项目链接？？
 - 交易 NFT，NFT 有一些属性，比如背景、武器，可以隐藏我的武器，有点买东西的时候尔虞我诈的过程。
 - 证明：确实我的 A 属性很强，但是不需要揭露 A 属性的真正数值。
 - 目的：让 NFT 好玩，比如在 Mint 或者交易的时候隐藏一些信息，类似交易集卡的小游戏。

 - zknft 的项目：
   - https://zknft.xyz/#/
   - 放到链下（或者 Layer 2），让开销更小。
   - 通过一个 simple zk proof 让主网确认


#### AA（Abstract Account）

Andrew ： https://github.com/PrivateGuardian-protocol/PrivateGuardian

**Reply by Insomnia501** [reply]
https://github.com/kevinz917/zk-NFT
在交易过程中，Alice可以证明自己的NFT的A属性很强，但是不需要公开自己的NFT的全部属性，其他买家可能会冲着A属性去购买（但是会有“该NFT的B属性很弱”的风险）
就是用zk的隐私能力丰富了NFT的玩法

**Reply by Insomnia501** [reply]
Hunter Z Hunter(A ZKML project)
https://app.buidlbox.io/projects/hunter-z-hunter
https://www.youtube.com/watch?v=k2nqjItJ8BU
寻宝游戏：系统给用户一些提示，用户来猜出这是什么物品，并拍照上传，如果猜对了可以获得奖励
比如系统给出的是“西瓜（一段话描述）”，我找到了一个西瓜（寻宝过程），然后拍照上传，当我提交一张图到smart contract上时，这张图送去链下做inference，如果inference出来那个结果是西瓜，也就是结果正确，此时会返回一个prove文件“确实经过了inference计算得到的结果是西瓜”并且智能合约收到prove后就会执行一个transaction来发放奖励给用户
